code/game/objects/items/weapons/storage/storage.dm

```diff
@@ -259,18 +259,12 @@
 		return 0 //Storage item is full
 
 	if(can_hold.len)
-		var/ok = 0
-		for(var/A in can_hold)
-			if(istype(W, A))
-				ok = 1
-				break
-		if(!ok)
+		if(!is_type_in_typecache(W, can_hold))
 			if(!stop_messages)
 				usr << "<span class='warning'>[src] cannot hold [W]!</span>"
 			return 0
-
-	for(var/A in cant_hold) //Check for specific items which this container can't hold.
-		if(istype(W, A))
+	else

```

Comment:
If you look at the comments in the cant_hold definition, it is only supposed to work when can_hold is empty, i was enforcing that. I'll swap it back if you want to, of course.